### ISSUE TITLE
match_strings(): Any way to return additional columns?

### ISSUE THREAD

**Issue by dbcandle** [Bergvca/string_grouper#8]
I can't tell if this is a design limitation, or a matter of my being a "Spamonista", rather than a "Pythonista" so please bear with me...
...I'm calling match_strings with 2 dataframes like this:

matches = match_strings(dfREFERENCE[dfREFERENCE_KeyColumn].squeeze(), dfSUBJECT[dfSUBJECT_KeyColumn].squeeze())

As you can see, since my dataframes contain more than 1 column, I specify the columns of interest (dfREFERENCE_KeyColumn & dfSUBJECT_KeyColumn).

Question: Is it fundamentally impossible to have matches return the other columns in the master & duplicates arguments to match_strings?  Once I match on a text column (I'm doing "fuzzy matching" or "record-linking" between datasets with a lot of columns), I want to retain all the other columns.

Again, I can't tell if this is a design limitation, or a matter of my being a "Spamonista", rather than a "Pythonista" so please bear with me...  while I'm a seasoned and aging IT guy, I mostly "push paperwork" these days, and am new to Python.

**Comment by dbcandle** [comment]
OK, maybe "RTFM" applies here.  After reading more on the "Code" tab here, I see an example that uses the StringGrouper class.  I *think* that is what I was looking for...
...fingers crossed!

**Comment by zhihongchensg** [comment]
@dbcandle  are you able to share. am also looking to return additional columns. 

thanks plenty

**Comment by dbcandle** [comment]
I am still struggling because I am new to Python, but see the examples listed on the github page.  Notice that the last example, which uses the underlying StringGrouper class, returns the column “Company CIK Key” in addition to the matched columns.  This seems to occur on the line:


companies['deduplicated_name'] = string_grouper.get_groups()

In addition to the github page at https://github.com/Bergvca/string_grouper be sure to see the other pages at https://bergvca.github.io/2020/01/02/string-grouper.html and https://bergvca.github.io/2017/10/14/super-fast-string-matching.html

Also look at the main source-code file: string_grouper.py - lots of cmments in the code there!

I think the trick is shown on the github page as mentioned above, but it may be necessary to tap into the indices matrix (see other pages above).

Basically, I have two dataframes I want to link on a string column, with both dataframes having several columns (5-10 columns each).  I want the result to include all 10-20 of the columns in the 2 matched dataframes, not just the 2 string columns matched.

If you learn any more, PLEASE let me know!  :-)

Phil

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by dbcandle** [comment]
In looking more, I just do not think the string_grouper package can return columns other than the 2 string columns being matched.

I think it is necessary to iterate through individual rows of one of the two dataframes, making a call to get_groups or get_match for each row.

Still experimenting though...

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by zhihongchensg** [comment]
am not sure, i suspect to get the additional column (or just the original index),  requires tingling with sparse_dot_topn.  whatever additional data required has to be "tagged along".  And sparse_dot_topn only stores the top x number of results so cant just retrieve the index. Maybe the easier solution is simply to do another round of matching using the result returned and append the additional columns required - in pd.

**Comment by dbcandle** [comment]
I think the trick is to use  “._matches_list” from the StringGrouper class.  I need to do some testing, but it looks like it returns the indexes and similarity after .fit() is run on the StringGrouper class.

Phil

**Comment by dbcandle** [comment]
I got it to work: including other columns, not just the similarity and two text columns being compared.

Source code enclosed.

  *   It initially pulls data from SQL Server, so you can ignore that.  Still, I enclose sql_server.py, should it amuse you.
  *   Rather than installing the string_grouper package, I added string_grouper.py to my project, and import from it directly.
  *   In the source code, I use the name dfREFERENCE rather than master, and dfSUBJECT, rather than duplicates.  For example, dfREFERENCE can be a published high-quality data-set.  dfSUBJECT is the “subject” of the comparison; messy data being matched against the “reference” data.
  *   The “core” of this – our initial dialog below - can be found starting on line 171. The key is to use string_grouper._matches_list after calling string_grouper.fit().  It is a simple matrix with the index of the 2 tables being compared, and the similarity.
  *   I ran this with about 1.1M records in dfREFERENCE, and 20K in dfSUBJECT:
     *   String_grouper.fit() ran in about 2 minutes.
     *   The main loop starting on line 198 ran in about 17 hours on a 5th generation i5 laptop, with good SSD and 16Gb RAM.

I’m new to Python.  In fact, other than VBA, I’ve not programmed much in recent years, but I was a programmer for many years starting in about 1980.  I switched to “paperwork” for the most part, but am still in “IT”.

I’ll be most interested in finding a way to make the main loop more efficient.  Not sure if I will “fiddle” with Python, or load the matrix back to SQL Server, and process there.

==> If you find anything grossly inefficient in the main Python loop starting on line 198, “I’m all ears”! 😊  <==


Phil

**Comment by Bergvca** [comment]
Hi @dbcandle,

I don't see the code attached. Would love to see what you are trying to do and if it can be improved. Perhaps it's functionality that can be added to the library?

**Comment by dbcandle** [comment]
I placed the 2 files I sent in the prior email on github: https://github.com/dbcandle/Test

Can you see that?

A few notes:

  *   Please don’t laugh at the code… I’m more of a paper-pusher these days (but was formerly a programmer since 1980).
  *   The code that does what I was struggling with is in the main file: “main refactor (SQL Server Hybrid v4.0).py”.  I also included “sql_server.py” should it amuse you (I’m an ex-Oracle DBA, and use SQL Server quite a bit).
  *   Rather than the terms “master” and “duplicates”, my own code uses the terms:
     *   dfSUBJECT – This is the “subject” of the searching, for example a messy list of software products (e.g. “Microzoft Office”).
     *   dfREFERENCE – This is what I’m comparing the “subject” data with, for example a very precise and accurate list of software products (e.g. “Microsoft Office 365”).
  *   The “main loop” starts on line 171.  It is based on:
     *   Calling string_grouper._matches_list directly, after calling string_grouper.fit().
     *   Using pandas to concatenate rows from dfREFERENCE and dfSUBJECT.  This allows returning any additional columns, not just the 2 columns used to generate string_grouper._matches_list.
  *   Again, please forgive my ignorance of Python.  For years, I’ve been looking for a fast and effective way to do “fuzzy matching” (a.k.a. record linking), but tools like OpenRefine and other free offerings were FAR TOO SLOW.  I suspect the reason that the “main loop” takes 17 hours for my data (dfSUBJECT has 20K rows, dfREFERENCE has 1.1M rows) is my use of pandas for each loop iteration.  I suspect that the main loop would be better using native Python data-types, but that is beyond my Python skill at this time.  I’m struggling to learn, simultaneously:
     *   scikit-learn
     *   pandas
     *   python
     *   git
     *   github
     *   Pycharm

Anyway, it is an honor that you would even contact me.  It looks like originally, another user of string_grouper contacted me, and CCed you, but I’m not sure.  Anyway…

May the source be with you! 😊


Phil

**Comment by Bergvca** [comment]
Hi @dbcandle,

It seems to me you want to join the reference table to the subject table using the output of match_strings, is that correct? So dfREFERENCE --> matches --> dfSUBJECT? If that is the case you can just use the *merge* or *join* function from pandas. 

Something like:

```python
matches = match_strings(dfREFERENCE[dfREFERENCE_KeyColumn], dfSUBJECT[dfSUBJECT_KeyColumn])
df = dfSUBJECT.merge(matches, left_on=dfREFERENCE_KeyColumn, right_on='left_side') # defaults to left join - not sure if that is what you want
df = df.merge(matches, left_on='right_side', right_on='dfSUBJECT_KeyColumn')
```

I did not test this, but it should give you the general idea.

**Comment by dbcandle** [comment]
Awesome!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by Bergvca** [comment]
@dbcandle - out of curiosity, did that suggestion help?

**Comment by dbcandle** [comment]
PERFECTLY!

I’ve been happily using string_grouper ever since!!!


Phil

**Comment by SidSaxena** [comment]
> matches = match_strings(dfREFERENCE[dfREFERENCE_KeyColumn], dfSUBJECT[dfSUBJECT_KeyColumn])
df = dfSUBJECT.merge(matches, left_on=dfREFERENCE_KeyColumn, right_on='left_side') # defaults to left join - not sure if that is what you want
df = df.merge(matches, left_on='right_side', right_on='dfSUBJECT_KeyColumn')


> PERFECTLY! I’ve been happily using string_grouper ever since!!! 

Did you have to change anything about this code snippet (besides the obvious names ofc) or did it run fine for you? Because I've tried to merge on different columns and different types (inner etc) but I still can't seem to get the desired columns. 

On running the above code I get an error on the  third statement. The first merge goes through just fine but it's the final merge where I'm encountering issues. If I switch up the left and right side I get a weirdly large dataframe with way too many columns and rows which is not the desired outcome of either the code or my intention.

Excuse me if this is a little dumb, I'm quite new to this stuff myself.

Thanks :)

**Comment by Bergvca** [comment]
This is now supported out of the box with the optional "id" variables.

**Comment by dbcandle** [comment]
A W E S O M E ! ! !
